include/igl/copyleft/cgal/triangulate.cpp

```diff
@@ -0,0 +1,143 @@
+#include "triangulate.h"
+#include "assign_scalar.h"
+#include "../../list_to_matrix.h"
+#include <CGAL/Point_2.h>
+#include <CGAL/Segment_2.h>
+#include <CGAL/Triangle_2.h>
+#include <CGAL/Constrained_Delaunay_triangulation_2.h>
+#include <CGAL/Constrained_triangulation_plus_2.h>
+#include <CGAL/Triangulation_face_base_with_info_2.h>
+#include <CGAL/Exact_predicates_exact_constructions_kernel.h>
+
+template <
+  typename DerivedV,
+  typename DerivedE,
+  typename DerivedH,
+  typename DerivedV2,
+  typename DerivedF2>
+IGL_INLINE void igl::copyleft::cgal::triangulate(
+  const Eigen::MatrixBase<DerivedV> & V,
+  const Eigen::MatrixBase<DerivedE> & E,
+  const Eigen::MatrixBase<DerivedH> & H,
+  const std::string flags,
+  Eigen::PlainObjectBase<DerivedV2> & TV,
+  Eigen::PlainObjectBase<DerivedF2> & TF)
+{
+  struct FaceInfo2
+  {
+    FaceInfo2(){}
+    bool visited = false;
+    bool in_domain = true;
+  };
+
+  typedef Eigen::Index Index;
+  typedef CGAL::Epeck Kernel;
+  typedef CGAL::Point_2<Kernel>    Point_2;
+  typedef CGAL::Segment_2<Kernel>  Segment_2;
+  typedef CGAL::Triangle_2<Kernel> Triangle_2;
+  typedef CGAL::Triangulation_vertex_base_2<Kernel>  TVB_2;
+  typedef CGAL::Triangulation_face_base_with_info_2<FaceInfo2,Kernel>    TFB_2;
+  typedef CGAL::Constrained_triangulation_face_base_2<Kernel,TFB_2> CTFB_2;
+  typedef CGAL::Triangulation_data_structure_2<TVB_2,CTFB_2> TDS_2;
+  typedef CGAL::Exact_intersections_tag Itag;
+  typedef CGAL::Constrained_Delaunay_triangulation_2<Kernel,TDS_2,Itag> CDT_2;
+  typedef CGAL::Constrained_triangulation_plus_2<CDT_2> CDT_plus_2;
+  typedef typename CDT_plus_2::Face_handle Face_handle;
+
+  CDT_plus_2 cdt;
+  // Insert all points
+  std::vector<Point_2> points;points.reserve(V.rows());
+  for(Eigen::Index i = 0;i<V.rows();i++)
+  {
+    Point_2 p(V(i,0),V(i,1));
+    cdt.insert(p);
+    points.emplace_back(p);
+  }
+  // Insert all edges 
+  for(Eigen::Index e = 0;e<E.rows();e++)
+  {
+    cdt.insert_constraint(points[E(e,0)],points[E(e,1)]);
+  }
+
+  // https://doc.cgal.org/latest/Triangulation_2/index.html#title30
+  //
+  // "remove" connected component of face start by marking faces as
+  // in_domain = false. start should not yet be visited.
+  const auto remove_cc = [&](Face_handle start)
+  {
+    std::list<Face_handle> queue;
```

Comment:
I wouldn't use a std::list for that. std::queue under the hood uses a std::dequeue, which should offer better performances.